Fixes and improvements for galaxy.json.

jmchilton opened the pull request:
- c715e9d fixes external metadata setting for these new primary datasets that I guess broke with tool working directory isolation added in Galaxy 16.04.
- 6c6574e fixes setting dbkey via galaxy.json which I guess broke years ago.
- 5437711 Enhances metadata setting allowed for explicit (as opposed to discovered) outputs - now has a test case and allows dbkey and info like option available for discovered datasets.
- 40ff8b0 Adds a test case demonstrating galaxy.json specification of metadata for discovered datasets.
- 1932fc1 contains a small improvement to reporting when testing metadata in tool tests.

Update:

I rebased a unit test fix, and pushed out two new commits - one that fixed a very old bug that wouldn't allow tools to 'only' test discovered datasets and one that added a new tool documenting specifying arbitrary metadata in outputs.

The failing framework test seems legitimate but runs fine locally, I wonder if the test framework has a race condition when using postgres.

jmchilton committed:
Improvements for metadata specification w/galaxy.json of explicit outputs.

I call explicit outputs galaxy.json entries of type="dataset" (as opposed to type="new_primary_dataset" - which I call discovered datasets). Previously galaxy.json could set the name and extension of these datasets. I've added a test case to verify this and a small tweak to the tool test framework to allow easy testing of such metadata setting.

In order to bring explicit output metadata setting closer to parity with discovered dataset metadata setting, I've added the ability to define 'dbkey' and 'info' entries in galaxy.json for these outputs.

jmchilton committed:
Fix setting dbkey in galaxy.json for discovered primary datasets.

Based on the change - I guess this must have been broken a long time ago - so maybe it isn't used? A subsequent change will add a test though - so hopefully there will be no regressions of this again.

jmchilton committed:
Fix setting metadata in galaxy.json for working directory isolation change.

This has been broken since 16.04 I guess? Maybe retry internally means it hasn't been an issue?

jmchilton committed:
Add tool test for discovered dataset metadata from galaxy.json.

jmchilton committed:
Improved error message when testing metadata in tool tests.

jmchilton committed:
Fix allow tool to just test discovered datasets.

Previously some sort of test on the base file the primary datasets were keyed on was required.

jmchilton committed:
Test tool demo-ing discovered dataset datatype metadata w/galaxy.json.

Also update the metadata test comparison to support non-textual values (e.g. floating point and integer metadata fields).

nsoranzo commented:
Failed framework test seems legit.

jmchilton committed:
Add hack to test sleeping when verifying metadata.

Seems like a race condition but I can't determine why there would be one here - this check should help determine if it is a race condition or something else.

jmchilton commented:
@nsoranzo Yeah - it was an hda id being conflated with a dataset id - this wasn't a problem in one-off testing but tested with the rest of the tests it is a problem I think - or maybe it was a type problem - either way I fixed both and verified the tool under question works with postgres now. Thanks for the catch.

jmchilton commented:
@bgruening and @nsoranzo - thanks for the reviews - either of you willing to merge?

mvdbeek commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1298,11 +1298,12 @@ def path_rewriter( path ):
+                        try:
+                            context_value = context[context_key]
+                            setattr(dataset, context_key, context_value)
+                        except Exception:
```
@jmchilton I know this was pre-existing, but is the expected Exception here only `KeyError` ?
Maybe sth. like

```
if context_key in context:
    context_value = context[context_key]
    setattr(dataset, context_key, context_value)
```

would be better ?

jmchilton replied:
It is an `ExpressionContext` not a dictionary - I just looked into the code and it looks like it would throw a `KeyError` and it would respect `in`. I'l make the change.

jmchilton replied:
Updated it - thanks for the suggestion.

jmchilton committed:
Rework small for loop per @mvdbeek's suggestion...

... see https://github.com/galaxyproject/galaxy/pull/2697#discussion_r74939815 for context.

mvdbeek commented:
Thanks @jmchilton, I'm sure this will come in handy !
